primitives/genesis-builder/src/lib.rs

```diff
@@ -0,0 +1,43 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+//! Substrate genesis config builder
+//!
+//! This Runtime API allows to construct `GenesisConfig`, in particular:
+//! - serialize the runtime default `GenesisConfig` struct into json format,
+//! - put the GenesisConfig struct into the storage. Internally this operation calls
+//!   `GenesisBuild::build` function
+//! for all runtime pallets, which is typically provided by pallet's author.
+//! - deserialize the GenesisConfig from given json blob and put GenesisConfig into the state
+//!   storage. Allows to build
+//! customized configuration.
+//!
+//! Providing externalities with empty storage and putting GenesisConfig into storage allows to
+//! catch and build the raw storage of `GenesisConfig` which is the foundation for genesis block.
+
+sp_api::decl_runtime_apis! {
+	/// API to interact with GenesisConfig for the runtime
+	pub trait GenesisBuilder {
+		/// Instantiate default `GenesisConfig` and serializes it to json blob.
+		fn default_genesis_config_as_json() -> sp_std::vec::Vec<u8>;
```

Comment:
```suggestion
		fn generate_as_json(name: sp_runtime::RuntimeString) -> sp_std::vec::Vec<u8>;
```

Naming is still not really good, but yeah... We also need to supply some kind of name or whatever we want to call it to distinguish which genesis config we want to generate. Here we also have different kind of genesis configs: https://github.com/paritytech/substrate/tree/2c3b923423fac829b02842fbb9a0016b55c417df/bin/node/cli/src/chain_spec.rs

Reply:
Maybe I am missing something, but I don't see the need for extra parameter to construct default `GenesisConfig`.
`GenesisConfig` of the runtime is just a struct containing all `GenesisConfig`s  pallets, so all required parametrization is already self-contained in config (AFAIU it).

What is see are different `ChainSpec`s: [development_config](https://github.com/paritytech/substrate/blob/17a99d11d2a7466dd43193cb9ecfdc9b441acd1d/bin/node/cli/src/chain_spec.rs#L388), [local_testnet_config](https://github.com/paritytech/substrate/blob/17a99d11d2a7466dd43193cb9ecfdc9b441acd1d/bin/node/cli/src/chain_spec.rs#L413).

Each of them has specific `genesis`: [local_testnet_genesis](https://github.com/paritytech/substrate/blob/17a99d11d2a7466dd43193cb9ecfdc9b441acd1d/bin/node/cli/src/chain_spec.rs#L403), [development_config_genesis](https://github.com/paritytech/substrate/blob/17a99d11d2a7466dd43193cb9ecfdc9b441acd1d/bin/node/cli/src/chain_spec.rs#L378). Those specific gensis are just differently parametrized `GenesisConfig` instances of `kitchensink_runtime` created in shared function [testnet_genesis](https://github.com/paritytech/substrate/blob/17a99d11d2a7466dd43193cb9ecfdc9b441acd1d/bin/node/cli/src/chain_spec.rs#L233).

With new `GenesisBuilder` API approach, that custom `GenesisConfig` will be embedded in the form of pre-configured json files customized on top of the `json` for default `GensisConfig`.

If I missed something, would you please be more precise on where the extra parameter is actually required?

Reply:
> With new `GenesisBuilder` API approach, that custom `GenesisConfig` will be embedded in the form of pre-configured json files customized on top of the `json` for default `GensisConfig`.

Could you give an example of this?

Reply:
BTW this `json` macro already exists: https://docs.rs/serde_json/latest/serde_json/macro.json.html

Reply:
Added in: 679a970e9

In my opinion we don't need to put this functionality into the runtime:
```
let mut json = Runtime::default_as_json();
merge(json, json!({
     "session": {
         "keys":  initial_authorities.map().collect(),
     },
      "staking": {
          "validator_count": validator_count,
      },
});
Runtime::build_from_json(json);
```

I am thinking of adding a small crate under `utils/genesis-builder` which could provide json merge function.

If additional dependency is not a problem we could use https://crates.io/crates/json-patch. But on the other hand, [patching json](https://github.com/idubrov/json-patch/blob/0e1f7aab17359d03bf662b3fd28c32a98e6d5958/src/lib.rs#L541) seems to be really simple, and maybe pulling the crate does not make sense.